### PR TITLE
feat(dev) Introduce settings to enable/disable dev data model elements

### PR DESCRIPTION
**Pull request by fpacifici**
When introducing a new dataset (or a new entity/storage in an existing dataset), we generally want to have the code in a functioning state before the schemas are applied and existing in all environments. We need to do development on the draft migrations before we have them applied to prod.
We do not have a good way to isolate parts of the data model that are still under development from the application in prod if we need to commit the code. 
(The DISABLED_DATASET field helps but it does not have the right granularity and it does not disable a storage set).

This introduces a flag in the settings that turns on/off features under development. Features under development can be datasets, entities and storages. These are declared as "dev" features in code in their factory and they would be invisible if the flag is off.
Of course this flag will be off in all environments except when a developer is relying on these features.

This change would not have been enough by itself to isolate a dataset (and solve the deployment issue we had with metrics). Storages are initialized when the module is loaded, thus in the current implementation they would try to instantiate a cluster for their storage set during initialization.

This is wrong. StorageSetKey is part of data model configuration, while the cluster config is specific to the environment, the data model config should always be the same everywhere, while the cluster config depends on the environment. So this also decouples the cluster initialization from the data model initialization. The data model initialization only stores the storage key now.

This would have still not been enough. To fully hide a new dataset we would also need to disable storage sets which are not assigned to a cluster yet. Thus the change to the clusters.py file.

### CODE REVIEW

File: snuba/clusters/cluster.py [hunk 1]
```diff
@@ -316,8 +312,12 @@ def __get_cluster_nodes(self, cluster_name: str) -> Sequence[ClickhouseNode]:
     _unique_registered_storage_sets
 ), "Storage set registered to more than one cluster"
 
+expected_storage_sets = {
+    s for s in StorageSetKey if s.value not in settings.DISABLED_STORAGE_SETS
+}
+
 assert (
-    set(StorageSetKey) == _unique_registered_storage_sets
+    not expected_storage_sets - _unique_registered_storage_sets
```
Comment (fpacifici) [inline snuba/clusters/cluster.py:320]: I think disabling a storage set while it is still associated to a cluster makes sense. No need to delete that config to disable it.

File: snuba/datasets/storage.py [hunk 2]
```diff
@@ -180,7 +180,7 @@ def __init__(
         )
         assert isinstance(schema, WritableTableSchema)
         self.__table_writer = TableWriter(
-            cluster=get_cluster(storage_set_key),
```
Comment (fpacifici) [inline snuba/datasets/storage.py:183]: This is where we were loading a cluster (environment specific config) while initializing the data model, which should be independent from the environment. I think we should avoid this practice and keep the two decoupled as much as possible during the initialization.

**Comment by fpacifici** [comment]
Now this hardcodes the development storage sets.
Also added a data model sanity test that loads all "non dev" datasets/entity/storage/storage_set to ensure the data model is sane.

File: tests/clusters/test_cluster.py [hunk 3]
```diff
@@ -1,50 +1,95 @@
...
+    "querylog",
+    "sessions",
+}
+
+ALL_STORAGE_SETS = {
+    "outcomes",
+    *ENABLED_STORAGE_SETS,
+}
+
+REDUCED_COPNFIG = [
```
Comment (lynnagara) [inline tests/clusters/test_cluster.py:27]: REDUCED_COPNFIG

**Review by lynnagara — APPROVED** [review]

**Pull request merged**